### PR TITLE
fix: enable limiting endowments with negative policy overrides

### PR DESCRIPTION
**Pull request by naugtur**

### CODE REVIEW

File: packages/core/test/mergePolicy.js [hunk 1]
```diff
@@ -48,6 +48,51 @@ testMerge('merge with resources', {
...
+        qwerty: false
+      }
+    }
+  }
+}, {
+  resources: {
+    babel: {
+      globals: {
+        abc: false,
+        'a.b.c': false, // this is not supported
```
Comment (naugtur) [inline packages/core/test/mergePolicy.js:71]: Should this work? I don't think that's currently possible.

**Comment by kumavis** [comment]
yeah i like this
can you add a failing test here https://github.com/LavaMoat/LavaMoat/blob/main/packages/core/test/globals.js

File: packages/core/test/mergePolicy.js (hunk 1 above)
Comment (kumavis) [reply packages/core/test/mergePolicy.js:71]: i'm not interested in supporting
```
`a.b` true
`a.b.c` false
```
at this time

**Review by weizman — APPROVED** [review]
pretty cool (just replace type `poperty`->`property`)

File: packages/core/src/makeGetEndowmentsForConfig.js [hunk 2]
```diff
@@ -30,36 +30,48 @@ function makeGetEndowmentsForConfig ({ createFunctionWrapper }) {
...
+      // false means no access. It's necessary so that overrides can also be used to tighten the policy
+      if (configValue === false) {
+        explicitlyBanned.push(path)
+        return 
+      }
       // write access handled elsewhere
       if (configValue === 'write') return
       if (configValue !== true) {
-        throw new Error(`LavaMoat - unknown value for config (${typeof configValue})`)
+        throw new Error(`LavaMoat - unrecognizable config value (${typeof configValue}) for path "${path}"`)
```
Comment (kumavis) [inline packages/core/src/makeGetEndowmentsForConfig.js]:
```suggestion
        throw new Error(`LavaMoat - unrecognizable policy value (${typeof configValue}) for path "${path}"`)
```

File: packages/core/src/makeGetEndowmentsForConfig.js [hunk 3]
```diff
@@ -30,36 +30,48 @@ function makeGetEndowmentsForConfig ({ createFunctionWrapper }) {
...
     paths.forEach(path => {
-      copyValueAtPath(path.split('.'), sourceRef, targetRef, unwrapTo, unwrapFrom)
+      copyValueAtPath('', path.split('.'), explicitlyBanned, sourceRef, targetRef, unwrapTo, unwrapFrom)
     })
     return targetRef
   }
 
-  function copyValueAtPath (pathParts, sourceRef, targetRef, unwrapTo = sourceRef, unwrapFrom = targetRef) {
+  function extendPath(visited, next){
+    if(!visited || visited.length === 0) return next
```
Comment (kumavis) [inline packages/core/src/makeGetEndowmentsForConfig.js]:
```suggestion
    if (!visited || visited.length === 0) return next
```

File: packages/core/src/makeGetEndowmentsForConfig.js [hunk 4]
```diff
@@ -30,36 +30,48 @@ function makeGetEndowmentsForConfig ({ createFunctionWrapper }) {
...
     const targetRef = {}
     paths.forEach(path => {
-      copyValueAtPath(path.split('.'), sourceRef, targetRef, unwrapTo, unwrapFrom)
+      copyValueAtPath('', path.split('.'), explicitlyBanned, sourceRef, targetRef, unwrapTo, unwrapFrom)
     })
     return targetRef
   }
 
-  function copyValueAtPath (pathParts, sourceRef, targetRef, unwrapTo = sourceRef, unwrapFrom = targetRef) {
+  function extendPath(visited, next){
```
Comment (kumavis) [inline packages/core/src/makeGetEndowmentsForConfig.js]:
```suggestion
  function extendPath(visited, next) {
```

File: packages/core/src/makeGetEndowmentsForConfig.js [hunk 5]
```diff
@@ -106,7 +118,13 @@ function makeGetEndowmentsForConfig ({ createFunctionWrapper }) {
         // the newly created container will be the next target
         nextTargetRef = containerRef
       }
-      copyValueAtPath(remainingParts, nextSourceRef, nextTargetRef)
+      copyValueAtPath(currentPath, remainingParts, explicitlyBanned, nextSourceRef, nextTargetRef)
+      return
+    }
+
+    // If conflicting rules exist, opt for the negative one. This should never happen
+    if(explicitlyBanned.includes(currentPath)) {
```
Comment (kumavis) [inline packages/core/src/makeGetEndowmentsForConfig.js]:
```suggestion
    if (explicitlyBanned.includes(currentPath)) {
```

File: packages/core/src/makeGetEndowmentsForConfig.js [hunk 6]
```diff
@@ -30,36 +30,48 @@ function makeGetEndowmentsForConfig ({ createFunctionWrapper }) {
...
+    if(!visited || visited.length === 0) return next
+    return `${visited}.${next}`
+  }
+
+  function copyValueAtPath (visitedPath, pathParts, explicitlyBanned, sourceRef, targetRef, unwrapTo = sourceRef, unwrapFrom = targetRef) {
     if (pathParts.length === 0) {
       throw new Error('unable to copy, must have pathParts, was empty')
     }
     const [nextPart, ...remainingParts] = pathParts
+    const currentPath = extendPath(visitedPath,nextPart)
```
Comment (kumavis) [inline packages/core/src/makeGetEndowmentsForConfig.js]:
```suggestion
    const currentPath = extendPath(visitedPath, nextPart)
```

**Review by kumavis — APPROVED** [review]

**Comment by kumavis** [comment]
nice

**Pull request merged**